src/script/lua_api/l_object.cpp

```diff
@@ -1701,43 +1701,183 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
 	return 1;
 }
 
-// set_sky(self, bgcolor, type, list, clouds = true)
+// set_sky(self, {bgcolor=, type=, param=, sky_color=, direc_fog, base_fog})
 int ObjectRef::l_set_sky(lua_State *L)
 {
 	NO_MAP_LOCK_REQUIRED;
 	ObjectRef *ref = checkobject(L, 1);
 	RemotePlayer *player = getplayer(ref);
-	if (player == NULL)
+	if (!player)
 		return 0;
 
-	video::SColor bgcolor(255,255,255,255);
-	read_color(L, 2, &bgcolor);
+	// Check for ColorSpec tables:
+	bool is_colorspec = false;
+	if (lua_istable(L, 2)) {
+		// Iterate over the whole table in case of missing ColorSpec keys:
+		// Only update the value if we know we have a valid ColorSpec table.
+		lua_getfield(L, 2, "r");
+		if (!is_colorspec) 
+			is_colorspec = lua_isnumber(L, -1);
+		lua_getfield(L, 2, "g");
+		if (!is_colorspec)
+			is_colorspec = lua_isnumber(L, -1);
+		lua_getfield(L, 2, "b");
+		if (!is_colorspec)
+			is_colorspec = lua_isnumber(L, -1);
+	}
+	
+	if (lua_istable(L, 2) && !is_colorspec) {
+		SkyboxParams skybox_params = player->getSkyParams();
 
-	std::string type = luaL_checkstring(L, 3);
+		lua_getfield(L, 2, "base_color");
+		if (!lua_isnil(L, -1))
+			read_color(L, -1, &skybox_params.bgcolor);
+		lua_pop(L, 1);
 
-	std::vector<std::string> params;
-	if (lua_istable(L, 4)) {
-		lua_pushnil(L);
-		while (lua_next(L, 4) != 0) {
-			// key at index -2 and value at index -1
-			if (lua_isstring(L, -1))
-				params.emplace_back(readParam<std::string>(L, -1));
-			else
-				params.emplace_back("");
-			// removes value, keeps key for next iteration
+		lua_getfield(L, 2, "type");
+		if (!lua_isnil(L, -1))
+			skybox_params.type = luaL_checkstring(L, -1);
+		lua_pop(L, 1);
+
+		lua_getfield(L, 2, "textures");
+		skybox_params.textures.clear();
+		if (!lua_istable(L, -1)) {
+		} else {
```

Comment:
`if (lua_istable(L, -1) && skybox_params.type == "skybox") {`
This would replace the empty case above and ignore the unused `textures` parameter for non-skybox modes.